pubsignals/common.go

```diff
@@ -0,0 +1,248 @@
+package pubsignals
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"reflect"
+
+	"github.com/iden3/go-circuits/v2"
+	parser "github.com/iden3/go-schema-processor/v2/json"
+	"github.com/iden3/go-schema-processor/v2/merklize"
+	"github.com/piprate/json-gold/ld"
+	"github.com/pkg/errors"
+)
+
+// todo: move & reuse to/from go-schem-processor
+
+// PropertyQuery struct
+type PropertyQuery struct {
+	FieldName     string
+	Operator      int
+	OperatorValue any
+}
+
+// QueryMetadata struct describe query metadata
+type QueryMetadata struct {
+	PropertyQuery
+	SlotIndex       int
+	Values          []*big.Int
+	Path            *merklize.Path
+	ClaimPathKey    *big.Int
+	Datatype        string
+	MerklizedSchema bool
+}
+
+const (
+	contextFullKey           = "@context"
+	serializationFullKey     = "iden3_serialization"
+	credentialSubjectFullKey = "https://www.w3.org/2018/credentials#credentialSubject" // #nosec G101
+)
+
+// ParseCredentialSubject parse credential subject and return array of property queries
+func ParseCredentialSubject(_ context.Context, credentialSubject any) (out []PropertyQuery, err error) {
+	if credentialSubject == nil {
+		return []PropertyQuery{
+			{
+				Operator:  circuits.EQ,
+				FieldName: "",
+			},
+		}, nil
+	}
+
+	out = []PropertyQuery{}
+
+	jsonObject, ok := credentialSubject.(map[string]interface{})
+	if !ok {
+		return nil, errors.New("Failed to convert credential subject to JSONObject")
+	}
+
+	entries := getObjectEntries(jsonObject)
+	if len(entries) == 0 {
```

Comment:
theoretically empty object is the same as
```if credentialSubject == nil {
		return []PropertyQuery{
			{
				Operator:  circuits.EQ,
				FieldName: "",
			},
		}, nil
	}
```

Reply:
done